### PR TITLE
calculator complete

### PR DESCRIPTION
**Pull request by alizaintern**
1)  All arithmetic function implemented.
2) All trigonometry  function implemented.
3) Operator precedence Handled.
4) Clear and backspace functions  implemented.
5) Variable declaration handled.

### CODE REVIEW

File: calculator.html [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        <button onclick="saveVariable()">Save Variable</button>
+       
+        
+     </div>
+     
+    </div>
+  </div>
+  
+  
+
```
Comment (prodcastorDev) [inline calculator.html]: Remove the unnecessary spaces

File: exprseeion.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	function prec(c) {
+		if(c == '^')
+			return 3;
+		else if(c == '/' || c=='*')
+			return 2;
+		else if(c == '+' || c == '-')
+			return 1;
+		else
+			return -1;
+	}
```
Comment (prodcastorDev) [inline exprseeion.js]: - Indentation is not correct
- Do not use ==, use ===
- Try to use early return:

`if (c === '^') return 3`

in this way you will not have to create multiple if else if conditions

File: exprseeion.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+	// The main function to convert infix expression
+	//to postfix expression
+	function infixToPostfix(s) {
+
+		let st = []; //For stack operations, we are using JavaScript built in stack
+		let result = "";
+
+		for(let i = 0; i < s.length; i++) {
```
Comment (prodcastorDev) [inline exprseeion.js]: use map or forEach, try not to use for loop in JS

File: exprseeion.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+				while(st.length != 0 && prec(s[i]) <= prec(st[st.length - 1])) {
+					result += st[st.length - 1];
+					st.pop();
+				}
+				st.push(c);
+			}
+		}
+
+		// Pop all the remaining elements from the stack
+		while(st.length != 0) {
```
Comment (prodcastorDev) [inline exprseeion.js]: `!st.length` is enough, != 0 is redundant

File: index.html [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+						onclick="base10Log()"></td>
+				<td><input type="button" value="l2e"
+						onclick="display.value += Math.LOG2E"></td>
+				<td><input type="button" value="l10e"
+						onclick="display.value += Math.LOG10E"></td>
+				<td><input type="button" value="exp("
+						onclick="display.value += 'exp('"></td>
+			</tr>
+		</table>
+	</form> -->
```
Comment (prodcastorDev) [inline index.html]: remove the commented out code

File: index.html [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        }
+	// function calculate() {
+	// 	try {
+	// 		const expression = document.getElementById('display').value;
+	// 		const result = math.evaluate(expression);
+	// 		document.getElementById('display').value = result;
+	// 	} catch (error) {
+	// 		document.getElementById('display').value = 'Error';
+	// 	}
+	// }
```
Comment (prodcastorDev) [inline index.html]: Remove the commented out code!

File: index.html [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+		expression+=value;
+		document.getElementById('result').textContent=expression;
+	}
+	function calculate() {
+            try {
+                let result = eval(expression); // Evaluate the expression
+                document.getElementById('result').textContent = result;
+                expression = result.toString(); // Update the expression with the result for further calculations
+            } catch (error) {
+                document.getElementById('result').textContent = 'Error';
```
Comment (prodcastorDev) [inline index.html]: why not clear the textContent and show error in a alert box!

File: index2.html [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+<form>
+		<table>
+			<tr>
+				<td colspan="6">
+					<input id="display" type="text">
+					<p id="result"></p>
+
+                
+				</td>
+			</tr>
```
Comment (prodcastorDev) [inline index2.html]: indentation is bad here!

File: index2.html [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+        //                 case 'Sqrt':
+        //                     operandStack.push(Math.pow(operand1, operand2));
+        //                     break;
+                        
+        //             }
+        //         }
+        //     }
+
+        //     return operandStack.pop();
+        // }
```
Comment (prodcastorDev) [inline index2.html]: remove the commented out code

File: index2.html [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+        
+        function isOperator(ch) {
+    return ['+', '-', '*', '/', 'sin', 'cos', 'tan', 'sqrt', '^'].includes(ch);
```
Comment (prodcastorDev) [inline index2.html]: Create a constant for `['+', '-', '*', '/', 'sin', 'cos', 'tan', 'sqrt', '^']`

File: script.js [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+      return 3;
+  
+    if (operator === '^') 
+      return 4;
+    
+    return 0;
+  }
+  
+  function isOperator(token) {
+    return ['+', '-', '*', '/', 'sin', 'cos', 'tan', 'sqrt','^'].includes(token);
```
Comment (prodcastorDev) [inline script.js]: create a constant and use it here as well!

File: script.js [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+          const operand2 = stack.pop();
+          const result = evaluateOperator(token, null, operand2);
+          stack.push(result);
+        }
+      }
+    }
+    const ans = document.getElementById('answer');
+    ans.value = stack.pop();
+    return ans.value;
+    //document.getElementById('answer').value=ans.value ;
```
Comment (prodcastorDev) [inline script.js]: remove commented out code

File: script.js [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+  const expression = expressionInput.value; // Extract the expression from the input field
+
+  // if (expression.includes('sin') || expression.includes('cos') || expression.includes('tan')) {
+  //   const postfix = infixToPostfix(expression); // Pass the expression to infixToPostfix()
+  //   const result = evaluatePostfix(postfix); // Assuming you have a function to evaluate postfix expressions
+  //   document.getElementById('answer').value = result;
+  // // } else {
+  //   const ans = eval(expression);
+  //   document.getElementById('answer').value = ans;
+  // }
```
Comment (prodcastorDev) [inline script.js]: Remove the commented out code!

File: script1.js [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+  }
+  
+  if( stack.length === 0)
+  {
+    return true;
+  } 
+  else{
+    return false;
+  }
+}
```
Comment (prodcastorDev) [inline script1.js]: indentation is not correct here

File: script1.js [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+
+
+
```
Comment (prodcastorDev) [inline script1.js]: remove extra spaces

File: script1.js [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+function radiansToDegrees(radians) {
+  return parseFloat(radians) * (180 / Math.PI);
+}
+
+
```
Comment (prodcastorDev) [inline script1.js]: remove extra spaces

File: script1.js [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+      
+    }
+    else if(i<=j)
+    {
+      continue;
+    }
+  else if (isOperator(tokens[i]))
+  {
+    temp=tokens[i];
+    //console.log(temp);
```
Comment (prodcastorDev) [inline script1.js]: remove commented out code

File: script1.js [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+
+  }
+  else if (isNumber(tokens[i]))
+  {
+    temp=tokens[i];
+    //console.log(temp);
```
Comment (prodcastorDev) [inline script1.js]: remove commented out code

File: script1.js [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+  }
+  else if( variablesDict.hasOwnProperty(tokens[i]))
+  {
+    temp= variablesDict[tokens[i]];
+    //console.log(temp);
```
Comment (prodcastorDev) [inline script1.js]: remove commented out code

File: test.js [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+// }
+
+
+// const expression = "+1+(+2)";
+// const result = removeUnaryPlus(expression);
+// console.log(result); // Output: "2 + 3 + 5"
+
+
+
+
```
Comment (prodcastorDev) [inline test.js]: Remove the commented out code

File: test.js [hunk 21]
```diff
@@ -0,0 +1,313 @@
...
+      if(tokens[i]=='Math.tan' && tokens[i+2]=='90')
+      {
+        return 'undefined ';
+      }
+      output.push(tokens[i]);
+      j=parseInt(i)+2;
+      
+       
+      temp=radiansToDegrees(tokens[j]);
+      //temp=parseFloat(tokens[j]);
```
Comment (prodcastorDev) [inline test.js]: remove commented out code

File: test.js [hunk 22]
```diff
@@ -0,0 +1,313 @@
...
+      
+    }
+    else if(i<=j)
+    {
+      continue;
+    }
+  else if (isOperator(tokens[i]))
+  {
+    temp=tokens[i];
+    //console.log(temp);
```
Comment (prodcastorDev) [inline test.js]: remove commented out code

File: test.js [hunk 23]
```diff
@@ -0,0 +1,313 @@
...
+
+  }
+  else if (isNumber(tokens[i]))
+  {
+    temp=tokens[i];
+    //console.log(temp);
```
Comment (prodcastorDev) [inline test.js]: remove commented out code

File: test.js [hunk 24]
```diff
@@ -0,0 +1,313 @@
...
+  }
+  else if(dict.hasOwnProperty(tokens[i]))
+  {
+    temp=dict[tokens[i]];
+    //console.log(temp);
```
Comment (prodcastorDev) [inline test.js]: remove commented out code

File: test.js [hunk 25]
```diff
@@ -0,0 +1,313 @@
...
+exp=exp.replace('cos','Math.cos');
+exp=exp.replace('tan','Math.tan');
+exp=exp.replace('e','Math.E');
+exp=exp.replace('pi','Math.PI');
+if(!isValidExpression(exp)){
+  console.log('not a valid expression ');
+  return;
+}
+const tokens = exp.match(/(?:sin|cos|tan|sqrt|\b[A-Za-z0-9.]{1,9}(?![\d.])\b|\^|\b[+\-*/()]\b|e|pi|((?:\d+(?:\.\d)?|\.\d+|[+\-*/()])))?/g);
+//console.log('Before changing' + tokens);
```
Comment (prodcastorDev) [inline test.js]: remove commented out code

File: test.js [hunk 26]
```diff
@@ -0,0 +1,313 @@
...
+let changedExpression=changeExpression(tokens,dict);
+if(changedExpression=='undfined')
+{
+  console.log('Undefined');
+}
+//console.log('After changing' + changedExpression);
```
Comment (prodcastorDev) [inline test.js]: remove commented out code

**Review by prodcastorDev — CHANGES_REQUESTED** [review]

File: calculator.html (hunk 1 above)
Comment (alizaintern) [reply calculator.html]: done , removed all the extra spaces

File: test.js (hunk 26 above)
Comment (alizaintern) [reply test.js]: sorry for inconvenience, but my project's files are Calculator.html, script1.js , style.css.  Rest are extra files, I deleted all of them now.

File: script1.js [hunk 27]
```diff
@@ -0,0 +1,170 @@
...
+          const lastOpeningBracket = stack.pop();
+          const correspondingOpeningBracket = openingBrackets[closingBrackets.indexOf(char)];
+          
+          if (lastOpeningBracket !== correspondingOpeningBracket) {
+              return false; // Mismatched brackets
+          }
+      }
+  }
+  
+  if( stack.length === 0)
```
Comment (prodcastorDev) [inline script1.js]: `!stack.length` is enough

File: script1.js [hunk 28]
```diff
@@ -0,0 +1,170 @@
...
+function isNumber(token) {
+  return !isNaN(parseFloat(token)) && isFinite(token);
+}
+
+function changeExpression(tokens)
+{
+let output=[];
+let temp='';
+let j;
+for(let i=0;i<tokens.length;i++ )
```
Comment (prodcastorDev) [inline script1.js]: please use map or forEach, try not to use for loop

File: script1.js (hunk 27 above)
Comment (alizaintern) [reply script1.js]: resolved

**Review by prodcastorDev — APPROVED** [review]

**Pull request merged**